servoshell: ohos / android: Decouple initialization and window creation

jschwe opened the pull request:
We can start initializing servo without needing to wait for the
creation of a window.
This allows us to perform initialization of servo itself,
concurrently to the app UI thread setting up the app UI.
Once the native window is ready, we can then let servo create
the first window and load the initial URL.
This is also interesting in the context of using servo as a
webview library, where loading the library / initialising the
webview and loading the first url are typically decoupled
steps.

Note: on the android port the Java code is not touched, which means that we effectively still 
perform the initialization at the same point in time on android. 
The change to call the base servo initialiation from the Java app, can be done in a seperate PR, perhaps by someone more familiar with android than me. 

Follow-up PRs will add multi-window support, which means that some of
the multi-webview related code in this PR has open todos,
which will be addressed by follow-ups.



Testing: The ohos-port is tested in CI, the android port was manually tested by me.

jschwe committed:
servoshell: ohos: Decouple initialization and window creation

We can start initializing servo without needing to wait for the
creation of a window.
This allows us to perform initalization of servo itself,
concurrently to the app UI thread setting up the app UI.
Once the native window is ready, we can then let servo create
the first window and load the initial URL.
This is also interesting in the context of using servo as a
webview library, where loading the library / initialising the
webview and loading the first url are typically decoupled
steps.

Follow-up PRs will add multi-window support, which means that some of
the multi-webview related code in this PR has open todos,
which will be addressed by follow-ups.

Signed-off-by: Jonathan Schwender <schwenderjonathan@gmail.com>

mrobinson commented on ports/servoshell/egl/app.rs:
```diff
@@ -258,7 +253,12 @@ pub(crate) struct AppInitOptions {
+    // todo: multi-window support, like desktop version.
+    // This is just an intermediate state, to split refactoring into
+    // multiple PRs.
+    // todo: active window?
```
Nit: typically `TODO` is spelled with capital letters in Servo, which really only matters when grepping.

The `RunningAppState` stores a list of windows which can give access to the `PlatformWindow` as well. Maybe that's enough here or do we need the actual type of the `PlatformWindow` implementation? I think this is stored as the `trait` in the desktop version. It might be nice to keep the same behavior between ports.

jschwe replied:
> Maybe that's enough here or do we need the actual type of the PlatformWindow implementation?

I think the trait currently doesn't offer access to the HostCallbacks / HostTrait, which are stored in the EmbeddedPlatformWindow. The desktop version doesn't have those, I guess because it assumes winit?

mrobinson replied:
I think we could get around this issue by accepting `Rc<HostTrait>` instead of `Box<HostTraits>` and storing it in both the `PlatformWindow` and the `App`. In fact we will have to do something like this if we want to support multiple windows for the embedded platform.

mrobinson commented on ports/servoshell/egl/ohos/mod.rs:
```diff
@@ -414,6 +395,42 @@ impl ServoAction {
+                // SET_URL_BAR_CB
+                //     .get()
+                //     .map(|f| f.call(url, ThreadsafeFunctionCallMode::Blocking));
+            },
```
Nit: I'd remove this dead code.

jschwe replied:
Thanks for catching that.

jschwe commented on ports/servoshell/egl/app.rs:
```diff
@@ -391,7 +427,16 @@ impl App {
+            //     .expect("resize called but platform window does not exist");
+            // self.state.window(ServoShellWindowId(0)).expect("Platform window doesn't exist").platform_window().
+            // *platform_window.viewport_rect.borrow_mut() = viewport_rect;
+        }
```
This is the problematic place, where we want access to the EmbeddedPlatformWindow. t modify the viewport_rect.

mrobinson replied:
I think a way to do this so that it would compatible with the desktop port is to expose the size of the window via `HostTrait` and then call `Window::rebuild_user_interface`. The `EmbeddedPlatformWindow` would then note that the inner size of the window had changed and update the `WebView` size. In fact, I think this is necessary in order to properly implement inner and outer size on the embedded platform.

mrobinson replied:
Okay. Now that #41773 has landed, I think you should be able to access the `EmbeddedPlatformWindow` when processing resizes.

jschwe committed:
fixup todo to uppercase

Signed-off-by: Jonathan Schwender <schwenderjonathan@gmail.com>

Narfinger commented on ports/servoshell/egl/app.rs:
```diff
@@ -292,25 +294,54 @@ impl App {
+            )
+            .expect("Could not create RenderingContext"),
+        );
+        if let Some(old) = self
```
This seems to contradict the naming a bit. Perhaps something like 'initialize_single_platform_window' or something that makes it clear it replaces it currently.

mrobinson commented on ports/servoshell/egl/app.rs:
```diff
@@ -292,25 +294,54 @@ impl App {
+        })
+    }
+
+    pub(crate) fn initialize_platform_window(
```
Let's get ready for multiwindow support by just calling this `add_platform_window`:

```suggestion
    pub(crate) fn add_platform_window(
```

mrobinson commented on ports/servoshell/egl/app.rs:
```diff
@@ -396,7 +431,10 @@ impl App {
-        *self.platform_window.viewport_rect.borrow_mut() = viewport_rect;
+        let window = self.window().platform_window();
+        let ohos_window = window.as_headed_window().expect("No headed window");
+        *ohos_window.viewport_rect.borrow_mut() = viewport_rect;
```
```suggestion
        let embedded_platform_window = window.as_headed_window().expect("No headed window");
        *embedded_platform_window.viewport_rect.borrow_mut() = viewport_rect;
```

mrobinson reviewed (changes requested):
Nice work, just a couple minor comments on the shared code. I don't feel confident enough to review the OHOS code, so I'll leave that for others.

jschwe committed:
don't store refreshdriver in the app

Signed-off-by: Jonathan Schwender <schwenderjonathan@gmail.com>

jschwe committed:
don't store RC in the app

Signed-off-by: Jonathan Schwender <schwenderjonathan@gmail.com>

jschwe committed:
Rename initialize_platform_window to add_platform_window

Signed-off-by: Jonathan Schwender <schwenderjonathan@gmail.com>

jschwe committed:
Apply review suggestions

Signed-off-by: Jonathan Schwender <schwenderjonathan@gmail.com>

jschwe committed:
Part 1: OpenHarmony initialization docs

Signed-off-by: Jonathan Schwender <schwenderjonathan@gmail.com>

mrobinson reviewed (approved):
The shared bits look totally fine to me. I'll let those more familiar with OHOS vouch for those bits.

jschwe commented on ports/servoshell/egl/ohos/mod.rs:
```diff
@@ -2,6 +2,47 @@
+/// code continues with `onWindowStageCreate()`, which will load our ArkUI page from
+/// `pages/Index.ets`
+///
+/// To be continued ...
```
Yeah, that's when I decided to head home. I'll continue with the rest of the initialization tomorrow.

jschwe commented on support/openharmony/entry/src/main/ets/entryability/EntryAbility.ets:
```diff
@@ -29,11 +30,11 @@ export default class EntryAbility extends UIAbility {
+          params.push(key)
+          if (value) {
+            params.push(value)
+          }
```
There is some minor spacing  inconsistency between the servodemo code and this repo. 
I synced the code from the demo to here, but probably the spacing should be 4 spaces, so I'll change that and sync it back to the servodemo PR

jschwe replied:
There is some minor spacing  inconsistency between the servodemo code and this repo. 
I synced the code from the demo to here, but probably the spacing should be 4 spaces, so I'll change that and sync it back to the servodemo PR

jschwe replied:
Looks like the DevEco Studio formats `ets` files with 2 spaces by default. I guess we should not change that, since otherwise the whole file would change.

jschwe commented on support/openharmony/entry/src/main/ets/pages/Index.ets:
```diff
@@ -1,24 +1,5 @@
-import deviceInfo from '@ohos.deviceInfo';
 import promptAction from '@ohos.promptAction';
-
-interface ServoXComponentInterface {
```
There are two ways to use a native shared library from napi. 
The standard way is by doing "import xxx from libservoshell.so" as we do after this PR. 
Xcomponent has a separate initialization mechanism and allows to access the methods on the xcomponent object (with a slightly different syntax to declare objects and functions).
Since we want to support multiple windows, we don't want to bind servo shell methods to xcomponent instances.

jschwe committed:
Finish documenting startup

Signed-off-by: Jonathan Schwender <schwenderjonathan@gmail.com>

jschwe committed:
Remove dead code in Index.ets

Signed-off-by: Jonathan Schwender <schwenderjonathan@gmail.com>

yezhizhen commented on ports/servoshell/egl/ohos/mod.rs:
```diff
@@ -517,24 +546,39 @@ fn main_thread(xc: XComponentWrapper, window: WindowWrapper) {
-        let _main_surface_thread = thread::spawn(move || {
-            main_thread(xc_wrapper, window_wrapper);
-        });
+    // TODO: This if will be removed once we add multi-window support in a follow-up PR.
```
```suggestion
    // TODO: This will be removed once we add multi-window support in a follow-up PR.
```

yezhizhen commented on ports/servoshell/egl/ohos/mod.rs:
```diff
@@ -517,24 +546,39 @@ fn main_thread(xc: XComponentWrapper, window: WindowWrapper) {
     } else {
         call(ServoAction::NewWebview(xc_wrapper, window_wrapper))
-            .expect("Could not create new webview");
+            .expect("Servo main thread channel not initialized");
```
```suggestion
            .expect("Could not create new webview");
```

jschwe replied:
Personally I think the new error message is better, since "main thread channel" not being there shows what the problem is quite obviously, while "Could not create new webview" doesn't really say anything about the why, and even sounds like it could be recoverable. The error here is not related to creating a webview at all, but really about not being able to reach the main thread even loop.

yezhizhen reviewed (approved):
Also remember to restore a few indents.

jschwe committed:
index.ets: restore indents

Signed-off-by: Jonathan Schwender <schwenderjonathan@gmail.com>

jschwe commented on ports/servoshell/egl/ohos/mod.rs:
```diff
@@ -517,24 +496,39 @@ fn main_thread(xc: XComponentWrapper, window: WindowWrapper) {
+
+        let native_vsync =
+            ohos_vsync::NativeVsync::new("ServoVsync").expect("Failed to create NativeVsync");
+        // get_period() returns an error - perhaps we need to wait until the first callback?
```
From the commented line directly below this one `native_vsync.get_period()`. 
I guess we could use this refactoring as an opportunity to drop that line and the comment.

jschwe committed:
Remove commented code.

We never further investigated this, so perhaps it's time to just delete
this commented info line.

Signed-off-by: Jonathan Schwender <schwenderjonathan@gmail.com>
